lib/m2ee/munin.py

```diff
@@ -174,55 +217,185 @@ def get_stats_from_runtime(client, config):
-    if type(stats['requests']) == list:
+    if type(stats["requests"]) == list:
         # convert back to normal, whraagh
         bork = {}
-        for x in stats['requests']:
-            bork[x['name']] = x['value']
-        stats['requests'] = bork
+        for x in stats["requests"]:
+            bork[x["name"]] = x["value"]
+        stats["requests"] = bork
 
     runtime_version = config.get_runtime_version()
     if runtime_version is not None and runtime_version >= 3.2:
         m2eeresponse = client.get_all_thread_stack_traces()
         if not m2eeresponse.has_error():
-            stats['threads'] = len(m2eeresponse.get_feedback())
+            stats["threads"] = len(m2eeresponse.get_feedback())
 
     java_version = guess_java_version(client, runtime_version, stats)
-    if 'memorypools' in stats['memory']:
-        memorypools = stats['memory']['memorypools']
+    return _populate_stats_by_java_version(stats, java_version), java_version
+
+
+def _populate_stats_by_java_version(stats, java_version):
+    # Notes from writing this function:
+    # Explanation of what each of these graphs mean:
+    # https://github.com/mendix/m2ee-tools/blob/master/doc/munin.md#the-smaps-statistics
+
+    # Explanation with some visuals (expand this comment)
+    # https://stackoverflow.com/questions/1262328/how-is-the-java-memory-pool-divided
+
+    # A memorypool is defined by the Jave docs as the following:
+    # "A memory pool represents the memory resource managed by the Java virtual
+    # machine and is managed by one or more memory managers."
+    # (source: https://docs.oracle.com/en/java/javase/11/docs/api/java.management/java/lang/management/MemoryPoolMXBean.html)  pylint: disable=line-too-long, noqa: E501
+
+    # The Mendix runtime uses the standard MemoryMXBean:
+    # The values from the `MemoryPoolMXBean`s go into the memorypool field and
+    # the values from the `MemoryMXBean` go into the init_, committed_, used_
+    # and max_ fields, both for the _heap and the nonheap memory.
+
+    # Since apparently in different Java versions these mean different things
+    # or are in different places, we need to shuffle these around depending
+    # on the Java version.
+
+    if "memorypools" in stats["memory"]:
+        standardized_memory_pools = _standardize_memory_pools_output(
+            stats["memory"]["memorypools"], java_version
+        )
+        stats["memory"].update(standardized_memory_pools)
+    elif java_version >= 8:
+        # We reach this branch only when we don't have the memorypools
+        # added, which is for Mendix runtime 6 and Java 8
+        memory = stats["memory"]
+        metaspace = memory["eden"]
+        eden = memory["tenured"]
+        survivor = memory["permanent"]
+        old = memory["used_heap"] - eden - survivor
+        memory["permanent"] = metaspace
+        memory["eden"] = eden
+        memory["survivor"] = survivor
+        memory["tenured"] = old
+    return stats
```

Comment:
Mendix 5.x runtimes (which runs on Java 7) are no longer supported by the buildpack. I tried deploying, but it gives a deprecation error.
So ideally we should never encounter a case where there are no `memorypools` and java version is less than 8.

But am not sure, if anyone could force the app to run on Java 7. Would that be possible ?

Reply:
No, that should not be possible, the customer would receive the same error that you did. (Nice that you checked by the way, I did not know what exactly would happen!)

Since you've found out that this branch should never be reached in Cloud V4, what do you think about just raising an exception here?

Reply:
I have made changes so that it raises exception. Also updated the tests to reflect the new behavior.

Exception in this case I feel is better, though its not backwards compatible, as it will fail sooner if ever it reaches the branch and alert us.

Reply:
Okay, I agree with your reasoning!